src/types/mapping.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use crate::err::{PyDowncastError, PyErr, PyResult};
+use crate::ffi;
+use crate::types::{PyAny, PySequence};
+use crate::AsPyPointer;
+use crate::{PyTryFrom, ToBorrowedObject};
+
+/// Represents a reference to a Python object supporting the mapping protocol.
+#[repr(transparent)]
+pub struct PyMapping(PyAny);
+pyobject_native_type_named!(PyMapping);
+pyobject_native_type_extract!(PyMapping);
+
+impl PyMapping {
+    /// Returns the number of objects in the mapping.
+    ///
+    /// This is equivalent to the Python expression `len(self)`.
+    #[inline]
+    pub fn len(&self) -> PyResult<usize> {
+        let v = unsafe { ffi::PyMapping_Size(self.as_ptr()) };
+        if v == -1 {
+            Err(PyErr::api_call_failed(self.py()))
+        } else {
+            Ok(v as usize)
+        }
+    }
+
+    /// Returns whether the mapping is empty.
+    #[inline]
+    pub fn is_empty(&self) -> PyResult<bool> {
+        self.len().map(|l| l == 0)
+    }
+
+    /// Determines if self contains the key `key`.
+    ///
+    /// This is equivalent to the Python expression `key in self`.
+    #[inline]
+    pub fn contains_key<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToBorrowedObject,
+    {
+        let r = key.with_borrowed_ptr(self.py(), |ptr| unsafe {
+            ffi::PyMapping_HasKey(self.as_ptr(), ptr)
```

Comment:
If we use `HasKey` which suppresses exceptions, the method can return just `bool`?

Otherwise, use `GetItem` internally to propagate the exceptions...